oxia-operator/controllers/oxiacluster_resources.go

```diff
@@ -0,0 +1,339 @@
+	_serviceAccount := &coreV1.ServiceAccount{
+		ObjectMeta: objectMeta(component, cluster),
+	}
+	if cluster.Spec.Image.PullSecrets != nil {
+		_serviceAccount.ImagePullSecrets = []coreV1.LocalObjectReference{{Name: *cluster.Spec.Image.PullSecrets}}
+	}
+	return _serviceAccount
+}
+
+func role(component Component, cluster *oxiav1alpha1.OxiaCluster) *rbacV1.Role {
+	if component == Coordinator {
+		return &rbacV1.Role{
+			ObjectMeta: objectMeta(component, cluster),
+			Rules: []rbacV1.PolicyRule{
+				//If storing shard state on the OxiaCluster status
+				policyRule("oxia.streamnative.io", []string{"oxiaclusters"}, []string{"get", "update"}),
+				//If storing shard state on a configmap data
+				policyRule("", []string{"configmaps"}, []string{"*"}),
+			},
+		}
+	}
+	return &rbacV1.Role{
+		ObjectMeta: objectMeta(component, cluster),
+		Rules:      []rbacV1.PolicyRule{},
+	}
+}
+
+func roleBinding(component Component, cluster *oxiav1alpha1.OxiaCluster) *rbacV1.RoleBinding {
+	_resourceName := resourceName(component, cluster.Name)
+	return &rbacV1.RoleBinding{
+		ObjectMeta: objectMeta(component, cluster),
+		Subjects: []rbacV1.Subject{{
+			Kind:      "ServiceAccount",
+			Name:      _resourceName,
+			Namespace: cluster.Namespace,
+		}},
+		RoleRef: rbacV1.RoleRef{
+			APIGroup: "",
+			Kind:     "Role",
+			Name:     _resourceName,
+		},
+	}
+}
+
+func service(component Component, cluster *oxiav1alpha1.OxiaCluster, ports []NamedPort) *coreV1.Service {
+	var clusterIp string
+	var publishNotReadyAddresses bool
+	if component == Server {
+		clusterIp = coreV1.ClusterIPNone
+		publishNotReadyAddresses = true
+	}
+	service := &coreV1.Service{
+		ObjectMeta: objectMeta(component, cluster),
+		Spec: coreV1.ServiceSpec{
+			Selector:                 selectorLabels(component, cluster.Name),
+			Ports:                    transform(ports, servicePort),
+			ClusterIP:                clusterIp,
+			PublishNotReadyAddresses: publishNotReadyAddresses,
+		},
+	}
+	service.Labels["oxia_cluster"] = cluster.Name
+	return service
+}
+
+func coordinatorConfigMap(cluster *oxiav1alpha1.OxiaCluster) (*coreV1.ConfigMap, error) {
+	servers := make([]ServerAddress, cluster.Spec.Server.Replicas)
+	for i := 0; i < int(cluster.Spec.Server.Replicas); i++ {
+		servers[i] = ServerAddress{
+			Public:   serviceAddress(cluster.Namespace, cluster.Name, i, PublicPort.Port),
+			Internal: fmt.Sprintf("%s-%d.%s:%d", cluster.Name, i, cluster.Name, InternalPort.Port),
+		}
+	}
+	config := ClusterConfig{
+		Namespaces: []NamespaceConfig{},
+		Servers:    servers,
+	}
+
+	for _, ns := range cluster.Spec.Namespaces {
+		config.Namespaces = append(config.Namespaces,
+			NamespaceConfig{
+				Name:              ns.Name,
+				InitialShardCount: ns.InitialShardCount,
+				ReplicationFactor: ns.ReplicationFactor,
+			})
+	}
+
+	bytes, err := yaml.Marshal(config)
+	return &coreV1.ConfigMap{
+		ObjectMeta: objectMeta(Coordinator, cluster),
+		Data: map[string]string{
```

Comment:
Will the configmap change when the configuration changes and should the service restart to load the new configuration, in which case you may need to configure configHash